protocol/block.go

```diff
@@ -63,58 +69,86 @@ func (c *Chain) GenerateBlock(ctx context.Context, prev *bc.Block, snapshot *sta
-			if err != nil {
-				return nil, nil, err
-			}
-			b.Transactions = append(b.Transactions, tx)
+		// Filter out double-spends etc.
+		err = newSnapshot.ApplyTx(tx.TxEntries)
+		if err != nil {
+			// TODO(bobg): log this?
+			continue
 		}
+
+		b.Transactions = append(b.Transactions, tx)
+		txEntries = append(txEntries, tx.TxEntries)
 	}
-	b.TransactionsMerkleRoot, err = bc.MerkleRoot(b.Transactions)
+
+	var err error
+
+	b.TransactionsMerkleRoot, err = bc.MerkleRoot(txEntries)
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "calculating tx merkle root")
 	}
-	b.AssetsMerkleRoot = result.Tree.RootHash()
-	return b, result, nil
+
+	b.AssetsMerkleRoot = newSnapshot.Tree.RootHash()
+
+	return b, newSnapshot, nil
+}
+
+// ValidateBlock validates an incoming block in advance of applying it
+// to a snapshot (with ApplyValidBlock) and committing it to the
+// blockchain (with CommitAppliedBlock).
+func (c *Chain) ValidateBlock(block, prev *bc.Block) error {
+	return c.validateBlock(block, prev, true)
+}
+
+func (c *Chain) validateBlock(block, prev *bc.Block, runProg bool) error {
+	var prevEntries *bc.BlockEntries
+	if prev != nil {
+		prevEntries = bc.MapBlock(prev)
+	}
+	err := validation.ValidateBlock(bc.MapBlock(block), prevEntries, c.InitialBlockHash, runProg)
+	if err != nil {
+		return errors.Sub(ErrBadBlock, err)
+	}
+	return nil
 }
 
-// ValidateBlock performs validation on an incoming block, in advance
-// of committing the block. ValidateBlock returns the state after
-// the block has been applied.
-func (c *Chain) ValidateBlock(ctx context.Context, prevState *state.Snapshot, prev, block *bc.Block) (*state.Snapshot, error) {
-	newState := state.Copy(prevState)
-	err := validation.ValidateBlockForAccept(ctx, newState, c.InitialBlockHash, prev, block, c.ValidateTxCached)
+// ApplyValidBlock creates an updated snapshot without validating the
+// block.
+func (c *Chain) ApplyValidBlock(block *bc.Block) (*state.Snapshot, error) {
+	newSnapshot := state.Copy(c.state.snapshot)
+	err := newSnapshot.ApplyBlock(bc.MapBlock(block))
 	if err != nil {
-		return nil, errors.Sub(ErrBadBlock, err)
+		return nil, err
+	}
+	if block.AssetsMerkleRoot != newSnapshot.Tree.RootHash() {
+		return nil, ErrBadStateRoot
 	}
-	// TODO(kr): consider calling CommitBlock here and
-	// renaming this function to AcceptBlock.
-	// See $CHAIN/protocol/doc/spec/validation.md#accept-block
-	// and the comment in validation/block.go:/ValidateBlock.
-	return newState, nil
+	return newSnapshot, nil
 }
 
-// CommitBlock commits the block to the blockchain.
+// CommitBlock commits a block to the blockchain. The block
+// must already have been applied with ApplyValidBlock or
+// ApplyNewBlock, which will have produced the new snapshot that's
+// required here.
 //
 // This function:
 //   * saves the block to the store.
-//   * saves the state tree to the store (optionally).
+//   * sometimes saves the state tree to the store.
 //   * executes all new-block callbacks.
```

Comment:
While you're updating these comments, "executes all new-block callbacks" is no longer true. That happens asynchronously through the block-processor pins.